pkg/storage/etcd/etcd_helper.go

```diff
@@ -42,64 +44,107 @@ import (
 )
 
 // storage.Config object for etcd.
-type EtcdConfig struct {
-	ServerList []string
-	Codec      runtime.Codec
-	Prefix     string
-	Quorum     bool
+type EtcdStorageConfig struct {
+	Config EtcdConfig
+	Codec  runtime.Codec
 }
 
 // implements storage.Config
-func (c *EtcdConfig) GetType() string {
+func (c *EtcdStorageConfig) GetType() string {
 	return "etcd"
 }
 
 // implements storage.Config
-func (c *EtcdConfig) NewStorage() (storage.Interface, error) {
-	cfg := etcd.Config{
-		Endpoints: c.ServerList,
-		// TODO: Determine if transport needs optimization
-		Transport: &http.Transport{
-			Proxy: http.ProxyFromEnvironment,
-			Dial: (&net.Dialer{
-				Timeout:   30 * time.Second,
-				KeepAlive: 30 * time.Second,
-			}).Dial,
-			TLSHandshakeTimeout: 10 * time.Second,
-			MaxIdleConnsPerHost: 500,
-		},
-	}
-	etcdClient, err := etcd.New(cfg)
+func (c *EtcdStorageConfig) NewStorage() (storage.Interface, error) {
+	etcdClient, err := etcd.New(c.Config.getEtcdClientConfig())
 	if err != nil {
 		return nil, err
 	}
-	return NewEtcdStorage(etcdClient, c.Codec, c.Prefix, c.Quorum), nil
+	return NewEtcdStorage(etcdClient, c.Codec, c.Config.Prefix, c.Config.Quorum), nil
+}
+
+// Configuration object for constructing etcd.Client
+// and accessing etcd.
+type EtcdConfig struct {
+	ServerList []string
+	Prefix     string
+	KeyFile    string
+	CertFile   string
+	CAFile     string
+	Quorum     bool
+}
+
+func (c *EtcdConfig) getEtcdClientConfig() etcd.Config {
+	return etcd.Config{
+		Endpoints: c.ServerList,
+		Transport: c.newHttpTransport(),
+	}
+}
+
+func (c *EtcdConfig) newHttpTransport() etcd.CancelableTransport {
+	var cc *tls.Config = nil
+	var err error
+
+	if c.KeyFile != "" && c.CertFile != "" {
+		var rpool *x509.CertPool = nil
+		if c.CAFile != "" {
+			if rpool, err = util.CertPoolFromFile(c.CAFile); err != nil {
+				glog.Errorf("Error reading Etcd CA File: %v", err)
+			}
+		}
+
+		if tlsCert, err := tls.LoadX509KeyPair(c.CertFile, c.KeyFile); err == nil {
+			cc = &tls.Config{
+				RootCAs:            rpool,
+				Certificates:       []tls.Certificate{tlsCert},
+				InsecureSkipVerify: true,
```

Comment:
previously, if a rootCA was specified, the connection was verified:
https://github.com/coreos/go-etcd/blob/master/etcd/client.go#L261-L296